File: code/dynamic_programming/rod_cutting/rod_cutting.go
package main

import  "fmt"


func max(x,y int) int {
	if x > y{
		return x
	}

	return y
}

func rod_cutting(prices []int) int {
	n := len(prices)
	max_prices := make([]int,n+1)
	max_prices[0] = 0
	var m int

	for i:=1;i<=n;i++{
		m = 0
		for j:=1;j<=i;j++{
			m = max(prices[j-1]+max_prices[i-j],m)
		}
		max_prices[i] = m
	}


	return max_prices[n]
}


func main(){

	prices := []int{1, 5, 8, 9, 10, 17, 17, 20}

	fmt.Println("Max possible value:",rod_cutting(prices))

}

File: cosmos-master/code/dynamic_programming/rod_cutting/rod_cutting.cpp
#include <iostream>
#include <vector>
// Part of Cosmos by OpenGenus Foundation
using namespace std;
int rodCutting(int rod_length, vector<int> prices){
    if (rod_length <= 0){
        return 0;
    }
    int maximum_price = INT8_MIN;
    for (int i = 0; i < rod_length; i++){
        maximum_price = max(maximum_price, prices[i] + rodCutting(rod_length-i-1, prices));
    }
    return maximum_price;
}
int main() {
    cout << rodCutting(8, {1,5,8,9,10,17,17,20}) << endl; // should be 22 (one length 2, another length 6)
    return 0;
}
